index.js

```diff
@@ -83,6 +93,106 @@ function attachToBody (options, req, reply, next) {
   })
 }
 
+function validateSerializedKey (tokens, value) {
+  let isValid = true
+
+  // first token can't be '[', ']', or '{}'
+  if (['[', ']', '{}'].includes(tokens[0])) isValid = false
+  // only a single {} special ending is allowed
+  else if (tokens.filter(t => t === '{}').length > 1) isValid = false
+  // {}, if present, must be the last token
+  else if (tokens.find(t => t === '{}') && tokens[tokens.length - 1] !== '{}') isValid = false
+
+  let hasOpenBracket = false
+  for (let i = 1; i < tokens.length; i++) {
+    const token = tokens[i]
+    const nextToken = tokens[i + 1]
+    // invalid closing bracket
+    if (!hasOpenBracket && token === ']') {
+      isValid = false
+      break
+    }
+    // invalid opening bracket
+    if (hasOpenBracket && token === '[') {
+      isValid = false
+      break
+    }
+    hasOpenBracket = token === '[' || (hasOpenBracket && token !== ']')
+    // orphan opening bracket
+    if (hasOpenBracket && nextToken === undefined) isValid = false
+    // {} is not allowed within brackets
+    if (hasOpenBracket && token === '{}') isValid = false
+    // after a closing bracket, only an opening bracket or {} are allowed
+    if (token === ']' && nextToken && !['[', '{}'].includes(nextToken)) {
+      isValid = false
+      break
+    }
+  }
+
+  // when the {} special ending is supplied, the value must be valid JSON
+  if (isValid && tokens[tokens.length - 1] === '{}') {
+    try {
+      secureJSON.parse(value)
+    } catch (e) {
+      return 'the {} special ending was used, but value is not a valid JSON'
+    }
+  }
+
+  if (tokens.includes('__proto__')) {
+    return '__proto__ is not allowed as a key name'
+  }
+
+  if (tokens.includes('constructor')) {
+    return 'constructor is not allowed as a key name'
+  }
+
+  return isValid ? null : 'invalid serialized key'
+}
+
+function getSerializedKeySegments (options, body, tokens, value) {
+  const segments = [{ parentRef: body, key: tokens[0] }]
+  const parseArrays = options.enableSpecialNotationKeys !== 'objectsOnly'
+
+  for (let i = 1; i < tokens.length; i++) {
+    const lastSegment = segments[segments.length - 1]
+    const token = tokens[i]
+    const prevToken = tokens[i - 1]
+    let parentRef
+    let key
+
+    if (token === ']' && prevToken === '[') {
+      parentRef = lastSegment.parentRef[lastSegment.key] ?? (parseArrays ? [] : {})
+      key = parseArrays ? parentRef.length : Object.keys(parentRef).length
+    } else if (parseArrays && token === ']' && /^\d+$/.test(prevToken)) {
```

Comment:
prevToken is a single character.
It is necessary to use regexp for this check?

Reply:
In this case, `prevToken` not necessarily will have a single character. This regex would also match a multi-digit number. But thanks to your comment, I remembered that I was not checking for leading zeroes, so I extracted this checking to a function and explicitly checked if the number has no leading zero. 

I've also [benchmarked this regex-based checking against two other alternatives](https://perf.link/#eyJpZCI6IndjcXoyZDM5OTFwIiwidGl0bGUiOiJDaGVja2luZyBpZiBzdHJpbmcgaXMgYSB2YWxpZCBpbnRlZ2VyIG51bWJlciA%2BPSAwIiwiYmVmb3JlIjoiY29uc3QgZGF0YSA9IFsnMCcsICcwMDEnLCAnMTAwJywgJzEyMzQnLCAnMjInLCAnOTknLCAnYWJjJywgJy0xMjMnLCAnMS43MicsICcxejMnXTtcblxuZnVuY3Rpb24gaXNWYWxpZEluZGV4VjEoc3RyKSB7XG4gIGlmIChzdHIubGVuZ3RoID4gMSAmJiBzdHJbMF0gPT09ICcwJykgcmV0dXJuIGZhbHNlO1xuICByZXR1cm4gL15cXGQrJC8udGVzdChzdHIpO1xufVxuXG5mdW5jdGlvbiBpc1ZhbGlkSW5kZXhWMihzdHIpIHtcbiAgaWYgKHN0ci5sZW5ndGggPiAxICYmIHN0clswXSA9PT0gJzAnKSByZXR1cm4gZmFsc2U7XG4gIGZvciAobGV0IGkgPSAwOyBpIDwgc3RyLmxlbmd0aDsgaSsrKSB7XG4gICAgY29uc3QgYyA9IHN0ci5jaGFyQ29kZUF0KGkpO1xuICAgIGlmIChjIDwgNDggfHwgYyA%2BIDU3KSByZXR1cm4gZmFsc2U7XG4gIH1cbiAgcmV0dXJuIHRydWU7XG59XG5cbmZ1bmN0aW9uIGlzVmFsaWRJbmRleFYzKHN0cikge1xuICBpZiAoc3RyWzBdID09PSAnLScpIHJldHVybiBmYWxzZTtcbiAgcmV0dXJuIHBhcnNlRmxvYXQoc3RyKS50b1N0cmluZygpID09PSBzdHIudG9TdHJpbmcoKTtcbn0iLCJ0ZXN0cyI6W3sibmFtZSI6IlJlZ2V4LWJhc2VkIiwiY29kZSI6ImRhdGEuZm9yRWFjaChzdHIgPT4gaXNWYWxpZEluZGV4VjEoc3RyKSkiLCJydW5zIjpbMjg2MDAwLDEzMDAwMCwxMDAwLDcyMDAwLDE1MTAwMCwxMDAwLDIyNjAwMCwzODAwMCwxODMwMDAsMjAxMDAwLDMyMTAwMCwyMjYwMDAsMTc4MDAwLDg0MDAwLDI5NTAwMCwxMDIwMDAsMzU1MDAwLDIxMDAwMCwzNzAwMCw5MzAwMCw4NzAwMCwyMzQwMDAsMjQ2MDAwLDIwMDAsMTAwMCwxMjMwMDAsMTk4MDAwLDE5ODAwMCwxNDAwMCwxMDAwLDEwMDAsMjAwMCwxOTcwMDAsNDYwMDAsMTAwMCwyMzMwMDAsMjI4MDAwLDMyMjAwMCwxODMwMDAsMTAxMDAwLDMyNjAwMCwyNDcwMDAsMjE0MDAwLDE4MTAwMCw2NTAwMCwxMDAwLDM1NTAwMCwxNjAwMCwxMDAwLDEwOTAwMCwyMTAwMCwzNTUwMDAsMTAwMCwyODAwMDAsMjA3MDAwLDIyMjAwMCwxMzkwMDAsMjIxMDAwLDMyMDAwLDExNDAwMCwyNTkwMDAsMTkwMDAwLDIzOTAwMCwxMTMwMDAsMTAwMCwyODUwMDAsMzQxMDAwLDI0NjAwMCw4MzAwMCwzMTIwMDAsMTUwMDAwLDE5MTAwMCwyMTcwMDAsMTkxMDAwLDEwMDAsOTMwMDAsMjQwMDAsMzE5MDAwLDIxNDAwMCwzNTUwMDAsMzU1MDAwLDQyMDAwLDEwMDAsMzcyMDAwLDEwMDAsNjgwMDAsMjg5MDAwLDEwMDAsNzEwMDAsMzEwMDAwLDIxOTAwMCw3OTAwMCwzNTUwMDAsMTAwMCwxMDAwLDI5NzAwMCwxMDAwLDI3OTAwMCwyNjUwMDAsMTc5MDAwXSwib3BzIjoxNTcyNTB9LHsibmFtZSI6Ikl0ZXJhdGlvbi1iYXNlZCIsImNvZGUiOiJkYXRhLmZvckVhY2goc3RyID0%2BIGlzVmFsaWRJbmRleFYyKHN0cikpIiwicnVucyI6WzI0MjAwMCwzNjAwMCwzMDAwLDE0NzAwMCwyOTAwMCwyNzMwMDAsMTUzMDAwLDYwMDAwLDgyMDAwLDI3NjAwMCwxMzQwMDAsMTQ1MDAwLDIyMDAwLDE0MTAwMCwzMTYwMDAsNjYwMDAsMjA4MDAwLDEyNzAwMCwxMzIwMDAsMTU4MDAwLDEzMjAwMCw5MjAwMCw0NDAwMCwxMDAwLDIzNjAwMCwxMjIwMDAsMTc3MDAwLDMyMzAwMCwxMzIwMDAsMzE4MDAwLDI3MjAwMCwzMTgwMDAsMTU4MDAwLDQ1MDAwLDEwMDAsMTAwMCwyNDUwMDAsODMwMDAsMTAwMCwxMTEwMDAsMjM5MDAwLDEwMDAsMTEzMDAwLDIwMDAwLDIxNDAwMCwyMzQwMDAsMjEzMDAwLDEzMjAwMCwxMzAwMCw0ODAwMCwyMjUwMDAsMTMyMDAwLDIxNjAwMCwxOTcwMDAsMTI3MDAwLDE2NjAwMCw2NDAwMCwxMzIwMDAsMjI4MDAwLDExMTAwMCwxMDcwMDAsMjIyMDAwLDk5MDAwLDIwMDAsMjE1MDAwLDIyNjAwMCwyNzAwMDAsMTU0MDAwLDI4MTAwMCwxMzIwMDAsNDQwMDAsMTY0MDAwLDEwMzAwMCwzMjAwMDAsMjUxMDAwLDkwMDAwLDIwNDAwMCwzMjAwMCw3MjAwMCwxOTkwMDAsMjA2MDAwLDE1NzAwMCwxMzgwMDAsMjAwMDAsMjQ3MDAwLDEzMjAwMCwzODAwMCwxMzUwMDAsMTQ2MDAwLDM1MDAwLDEzMjAwMCw1MzAwMCwxOTAwMCwxMDAwLDEzMjAwMCwxMjYwMDAsMjQxMDAwLDIyNjAwMCwxNzcwMDAsMTMwMDAwXSwib3BzIjoxNDA2NTB9LHsibmFtZSI6InBhcnNlRmxvYXQtYmFzZWQiLCJjb2RlIjoiZGF0YS5mb3JFYWNoKHN0ciA9PiBpc1ZhbGlkSW5kZXhWMyhzdHIpKSIsInJ1bnMiOlsyMzYwMDAsMTAwMCwxMDAwLDMxODAwMCwxMjQwMDAsMTgxMDAwLDEwMDAsMjM0MDAwLDI3NzAwMCwyNjUwMDAsMTk4MDAwLDY5MDAwLDcwMDAsMzU2MDAwLDE4MzAwMCwzNDAwMCwzNjIwMDAsMjAwMCw2MDAwLDI0NTAwMCwzNTAwMCwyNDYwMDAsMzMwMDAsMzAzMDAwLDIzNjAwMCwyOTIwMDAsMjExMDAwLDE3MDAwMCwxMDAwLDcwMDAsMjU0MDAwLDI2NTAwMCwzMzEwMDAsMjQ1MDAwLDEwMDAsMjgxMDAwLDE4NDAwMCwxOTYwMDAsMTAwMCw4NDAwMCwxMDAwLDIxOTAwMCwxNjYwMDAsMTAwMCwxMTIwMDAsNDYwMDAsMTIxMDAwLDE3MzAwMCwxODIwMDAsMjI3MDAwLDY1MDAwLDIyMjAwMCwxMDAwLDE3NDAwMCwzMjAwMDAsMjU1MDAwLDE3NDAwMCwyNzcwMDAsMjE1MDAwLDEwMDAsMTAwMCwxODcwMDAsMjczMDAwLDEwMDAsMjQ5MDAwLDE2MjAwMCwyNjIwMDAsOTAwMCwxMDAwLDEyMjAwMCwxMDAwLDEwMDAsMTc0MDAwLDIzMDAwLDEwMDAsMzA4MDAwLDI5MDAwLDEwMDAsMTYwMDAwLDI0ODAwMCwyOTQwMDAsMTAwMCwxNDUwMDAsNjgwMDAsODYwMDAsNTEwMDAsMjkwMDAsMjYzMDAwLDIzMDAwMCwxNzMwMDAsMjI2MDAwLDMzMTAwMCwyMDAwMDAsMTAxMDAwLDc4MDAwLDEwMDAsMjc5MDAwLDE0NTAwMCwxNzQwMDAsMjMwMDAwXSwib3BzIjoxNDc0NjB9XSwidXBkYXRlZCI6IjIwMjMtMDUtMDlUMTg6NDY6MTIuNTkyWiJ9), one iteration-based and other that relies on parseFloat. The regex version won by a slim margin.

Reply:
Unless I read the code wrongly `const prevToken = tokens[i - 1]`.
`prevToken` is never reassigned and is always a single character.
That means matching multi-digit number is useless for `prevToken`.

Reply:
`tokens` is not a string, it's an array of strings. So, `prevToken` might be either `[`, `]`, `{}`, a multi-digit numeric index or an alphanumeric object key. Besides, `prevToken` gets reassigned on each iteration.

Reply:
Then it doesn't make sense why doesn't use a single for-loop.

Reply:
The logic I've proposed relies on more than a loop in order to separate concerns. Mixing all steps in a single loop would make it way harder to read, debug and maintain.